site/posts/templating-isnt-just-for-web-developers.md

```diff
@@ -0,0 +1,408 @@
+
+That's it.  You should now be set up with an environment that
+has (at least) python3 and jinja2 installed in it.
+
+## The Use Case
+
+Imagine that you are the attorney for an oil tycoon with $4.5M
+held in the Bank of [Freedonia](https://en.wikipedia.org/wiki/Duck_Soup_(1933_film)).
+With tensions mounting between Freedonia and neighboring Sylvania,
+and war looming on the horizon, your client has authorized you to
+move these assets to another bank outside of Freedonia.  Clearly,
+your only course of action is to email as many people for whom
+you can find addresses to ask them if they would be willing to
+hold the assets in their account for the duration of the conflict.
+Obviously, you will need to pay them for their kind services, and
+the oil tycoon has agreed to authorize you to gift 20% of the assets
+to whomever holds them, as payment for their kindness.  You just
+need some personal information from them and some bank routine
+information.  That's all.  You have obtained your list of addressees
+from a kind person working for the humanitarian WikiLeaks "Foundation,"
+and now you must craft an email to send to all of these addressees.
+
+So, there is your entirely believable, totally serious, and extremely
+common scenario motivating your need to learn templating.
+
+## The Letter
+
+Fortunately, you already have a "template" for a formal *written* business
+letter sitting around on your hard drive.  It looks like this:
+
+```text
+[Your Address]
+
+[Date]
+
+[Recipient Name]
+[Recipient Title]
+[Recipient Company]
+[Recipient Address]
+
+[Greeting] [Recipient Name]:
+
+[Letter Body]
+
+[Closing],
+
+[Your Name]
+[Your Title]
+```
+
+Each element in this "template" that is enclosed in `[]`-brackets is
+just a placeholder for some other text.  Some of that text is short,
+like names, titles, or greeting or closing text.  Some of that text
+will me multi-line input, such as the letter's body and the addresses.
+In Python, all you would want to do is something like string
+substitution, for example replacing `[Your Name]` with `Jane Smith`.
+
+This is exactly what "templating" is all about.  It is, essentially,
+just string substitution...but with a touch more sophistication.
+
+## The Jinja2 Template
+
+So, to start out, we are going to convert this text template into a
+Jinja2 template.  To do so, the easiest first step would be to just
+replace every `[]`-bracketted item with a Jinja2 *expressions*,
+which are repressented with `{{}}`-brackets:
+
+```jinja
+{{ your_address }}
+
+{{ date }}
+
+{{ recipient_name }}
+{{ recipient_title }}
+{{ recipient_company }}
+{{ recipient_address }}
+
+{{ greeting }} {{ recipient_name }}:
+
+{{ letter_body }}
+
+{{ closing }},
+
+{{ your_name }}
+{{ your_title }}
+```
+
+Also, notice that I have converted the text labels inside the
+`[]`-brackets into valid Python variable names.  Let's assume
+that this new *template file* is called `letter.j2` (where
+the `.j2` means it is a Jinja2 template).
```

Comment:
The `.j2` file extension is not a must, right?

Reply:
Correct.  As far as I know, only the VS Code Jinja2 extension cares about the actual extension.  Jinja2 does not care at all.  I'll commend on that.

```suggestion
that this new *template file* is called `letter.j2` (where
the `.j2` means it is a Jinja2 template, but Jinja2 does not
care what extension the file actually has).
```